### ISSUE TITLE
Integrating audio.whisper and audio.vadsilero

### ISSUE THREAD

**Issue by jmgirard** [bnosac/audio.whisper#62]
Does this look correct for what `predict.whisper()` is looking for in its `segments` argument?

```r
convert_silero <- function(infile) {
  # Read in output from audio.vadsilero::silero()
  sdf <- readRDS(infile)
  # Extract segments information
  sections <- sdf$vad_segments
  # Drop non-voiced segments
  out <- sections[sections$has_voice == TRUE, ]
  # Convert to milliseconds
  out$start <- out$start * 1000
  out$end <- out$end * 1000
  # Calculate duration
  out$duration <- out$end - out$start
  # Drop segments with no duration
  out <- out[out$duration > 0, ]
  # Format output
  out <- out[, c("start", "duration")]

  out
}
```

**Comment by jwijffels** [comment]
Yes, something like that. I´ve made a generic function is.voiced which is part of audio.vadwebrtc which allows you to be a bit more liberal on small sections e.g. consider non-voiced segments smaller than 1 sec as voiced - we can be a bit liberal on identifying voiced segments and need to remove solely the larger chunks of silences to exclude non-voiced hallucinations.
See https://github.com/bnosac/audio.vadwebrtc/blob/0ca8192268b74f37a5b068775bad2f596cd339d5/R/vad.R#L133

**Comment by jwijffels** [comment]
Probably the VAD can be used for better diarization as well.

**Comment by jmgirard** [comment]
Ok, good to know. I'll try `offset` and `duration` instead.

**Comment by jmgirard** [comment]
Hmm... `offset` and `duration` seems to be running a lot slower than `sections`. Has that been your experience too?

**Comment by jwijffels** [comment]
For this you need to understand that whisper runs in chunks of 30seconds.

The because behaviour is different for the 2 arguments.

- argument sections creates a new audio file based on these voiced sections and does the transcription 
- arguments offset/duration (which will also look a bit around the cutoff timepoints) looks to each section, gets the 30 seconds where it is in, does the transcription of that and limits the output to the requested period. So this might do things several times if you have many voiced sections within the same 30 seconds window.

**Comment by jwijffels** [comment]
Would be good if you can test if the timepoints on the output when using sections are ok. 
Regarding speed that's normal. With sections you basically remove the non-voiced audio (so it will be faster than transcribing the full audio file). Probably you feeded the VAD directly in there but the VAD can provide many small chunks, it makes sense to combine these a bit. Function is.voiced which is part of audio.vadwebrtc (and works on output of  silero as well)  combines a bit larger chunks - https://github.com/bnosac/audio.vadwebrtc/blob/master/R/vad.R#L126-L184

**Comment by jmgirard** [comment]
Ok, `is.voiced()` is useful. Trying this now:

```r
convert_silero <- function(infile) {
  # Read in output from audio.vadsilero::silero()
  sdf <- readRDS(infile)
  # Extract segments information
  sections <- audio.vadwebrtc::is.voiced(sdf, units = "milliseconds", silence_min = 500, voiced_min = 500)
  # Drop non-voiced segments
  out <- sections[sections$has_voice == TRUE, ]
  # Format output
  out[, c("start", "duration")]
}

transcribe_file <- function(infile, outfile, vadfile) {
  if (file.exists(outfile)) {
    return("skipped")
  }
  vad_segments <- convert_silero(vadfile)
  transcript <- predict(
    model,
    infile,
    type = "transcribe",
    language = "en",
    n_threads = 1,
    n_processors = 1,
    #sections = vad_segments,
    offset = vad_segments$start,
    duration = vad_segments$duration,
    trace = FALSE
  )
  write_rds(transcript, file = outfile, compress = "gz")
  return("created")
}
```

**Comment by jwijffels** [comment]
> Probably the VAD can be used for better diarization as well if we do VAD by channel and see which section in the transcription corresponds to voiced elements as detected by the VAD

Added predict.whisper_transcription for this in version 0.4.1

**Comment by jwijffels** [comment]
Closing as the upgrade https://github.com/bnosac/audio.whisper/releases/tag/0.5.0 allows to integrate the VAD directly (using Silero)
